In [26]:
import pandas as pd
import numpy as np
import statsmodels.api as sm

In [27]:
df = pd.read_excel('/Users/fengyujie/Desktop/hw6.xlsx')
print(df)

          Date      SPX TR  3M Govt  Div Yield  Real Dividends
0          NaT     12.2531   0.0200   0.044580       10.343730
1   1940-02-29     12.3344   0.0200   0.045622       10.522553
2   1940-03-29     12.4565   0.0200   0.046639       10.688565
3   1940-04-30     12.4667   0.0200   0.046930       10.854539
4   1940-05-31      9.4804   0.1000   0.055319       11.019718
..         ...         ...      ...        ...             ...
967 2020-08-31  66127.7115   0.0938   0.017285       59.757033
968 2020-09-30  63615.0711   0.0915   0.017476       59.886866
969 2020-10-30  61923.3039   0.0864   0.017244       59.957888
970 2020-11-30  68701.5448   0.0712   0.016637       59.993094
971 2020-12-31  71340.4220   0.0583   0.015991       59.983781

[972 rows x 5 columns]


# Variance Ratio Test

In [28]:
df = df.rename(columns={"Date": "Date", "SPX TR": "spx_tr", 
                   "3M Govt": "3m_govt", "Div Yield": "div_yield", "Real Dividends": "real_div"})
df.head()


Date   spx_tr  3m_govt  div_yield   real_div
0        NaT  12.2531     0.02   0.044580  10.343730
1 1940-02-29  12.3344     0.02   0.045622  10.522553
2 1940-03-29  12.4565     0.02   0.046639  10.688565
3 1940-04-30  12.4667     0.02   0.046930  10.854539
4 1940-05-31   9.4804     0.10   0.055319  11.019718

In [29]:
df['lg_spx_tr'] = np.log(df['spx_tr'])
df['lg_3m_govt'] = np.log(df['3m_govt'])
df['lg_excess_return'] = df['lg_spx_tr'] - df['lg_3m_govt']
df.head()


/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/pandas/core/arraylike.py:397: RuntimeWarning: divide by zero encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)
/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/pandas/core/arraylike.py:397: RuntimeWarning: invalid value encountered in log
  result = getattr(ufunc, method)(*inputs, **kwargs)


Date   spx_tr  3m_govt  div_yield   real_div  lg_spx_tr  lg_3m_govt  \
0        NaT  12.2531     0.02   0.044580  10.343730   2.505779   -3.912023   
1 1940-02-29  12.3344     0.02   0.045622  10.522553   2.512392   -3.912023   
2 1940-03-29  12.4565     0.02   0.046639  10.688565   2.522243   -3.912023   
3 1940-04-30  12.4667     0.02   0.046930  10.854539   2.523061   -3.912023   
4 1940-05-31   9.4804     0.10   0.055319  11.019718   2.249227   -2.302585   

   lg_excess_return  
0          6.417802  
1          6.424415  
2          6.434266  
3          6.435084  
4          4.551812

In [30]:
var1 = statistics.variance(df['lg_excess_return'].loc[99:110])
var2 = statistics.variance(df['lg_excess_return'].loc[99:122])
var3 = statistics.variance(df['lg_excess_return'].loc[99:134])
var4 = statistics.variance(df['lg_excess_return'].loc[99:146])
var5 = statistics.variance(df['lg_excess_return'].loc[99:158])
var6 = statistics.variance(df['lg_excess_return'].loc[99:170])
var7 = statistics.variance(df['lg_excess_return'].loc[99:182])
var8 = statistics.variance(df['lg_excess_return'].loc[99:194])
var9 = statistics.variance(df['lg_excess_return'].loc[99:206])
var10 = statistics.variance(df['lg_excess_return'].loc[99:218])
variance = [var1, var2, var3, var4, var5, var6, var7, var8, var9, var10]

variance_ratio_output = pd.DataFrame()

for t in np.arange(0,10):
    variance_ratio = variance[t]/((t+1)*var1)
    variance_ratio_output[t] = [round(variance_ratio, 6)]

variance_ratio_output = variance_ratio_output.rename(columns= {0:"1948", 1:"1949", 2:"1950", 3:"1951", 4:"1952", 5:"1953", 6:"1954",7:"1955",8:"1956", 9:"1957"},
                            index={0: "variance ratio"})

variance_ratio_output


d = {'variance ratio': [1.0, 0.901279, 0.838306, 0.505766, 3.212217, 2.185096]}

variance_ratio_test = pd.DataFrame(data = d, index = ['1948', '1949', '1950', '1952', '1954', '1957'])

variance_ratio_test

variance ratio
1948        1.000000
1949        0.901279
1950        0.838306
1952        0.505766
1954        3.212217
1957        2.185096

# Test Statistical Significance 

In [31]:
n = len(variance)
test_stat = pd.DataFrame()
for t in np.arange(0,10):
    jrq = variance[t]/((t+1)*var1) -1
    z_stat = (jrq/((2*((t+1)-1))**0.5))*(n**0.5)
    test_stat[t] = [round(z_stat, 6)]
test_stat = test_stat.rename(index= {0: 'z statistics'})
test_stat



/var/folders/ry/0pcs2mmn24g1hyj6q9gm05lr0000gn/T/ipykernel_23447/4043931109.py:5: RuntimeWarning: invalid value encountered in double_scalars
  z_stat = (jrq/((2*((t+1)-1))**0.5))*(n**0.5)


0         1        2         3        4         5         6  \
z statistics NaN -0.220747 -0.25566 -0.410915 -0.55257  0.034391  2.019468   

                     7         8         9  
z statistics  1.616257  1.088561  0.883318

In [32]:

for t in np.arange(0,10):
    if (test_stat[t] > 0.97725).bool():
        print('reject the null')
    else:
        print('do not reject the null')
        

do not reject the null
do not reject the null
do not reject the null
do not reject the null
do not reject the null
do not reject the null
reject the null
reject the null
reject the null
do not reject the null


     For vairance ratio calculated in problem 1:
     Horizon of 1: 1948 --- Do not reject the null, indicating there is no mean reversion;
     Horizon of 2: 1949 --- Do not reject the null, indicating there is no mean reversion;
     Horizon of 3: 1950 --- Do not reject the null, indicating there is no mean reversion;
     Horizon of 5: 1952 --- Do not reject the null, indicating there is no mean reversion;
     Horizon of 7: 1954 --- Reject the null, indicating there is mean reversion;
     Horizon of 10: 1957 --- Do not reject the null, indicating there is no mean-reversion.

# Equity Return / Dividend Growth Predictability


## Regress $R_{t,\ t+k}$

### Year One

In [44]:
x = df['div_yield'].iloc[99:110].tolist()
y = df['lg_excess_return'].iloc[99:110].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.886
Model:                            OLS   Adj. R-squared:                  0.873
Method:                 Least Squares   F-statistic:                     69.77
Date:                Thu, 27 Oct 2022   Prob (F-statistic):           1.57e-05
Time:                        17:26:48   Log-Likelihood:                 23.085
No. Observations:                  11   AIC:                            -42.17
Df Residuals:                       9   BIC:                            -41.37
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.5560      0.168     27.190      0.0

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year Two

In [45]:
x = df['div_yield'].iloc[111:122].tolist()
y = df['lg_excess_return'].iloc[111:122].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.163
Model:                            OLS   Adj. R-squared:                  0.070
Method:                 Least Squares   F-statistic:                     1.753
Date:                Thu, 27 Oct 2022   Prob (F-statistic):              0.218
Time:                        17:27:03   Log-Likelihood:                 8.0296
No. Observations:                  11   AIC:                            -12.06
Df Residuals:                       9   BIC:                            -11.26
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          5.3309      1.585      3.364      0.0

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year Three

In [46]:
x = df['div_yield'].iloc[123:134].tolist()
y = df['lg_excess_return'].iloc[123:134].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.200
Model:                            OLS   Adj. R-squared:                  0.111
Method:                 Least Squares   F-statistic:                     2.245
Date:                Thu, 27 Oct 2022   Prob (F-statistic):              0.168
Time:                        17:27:18   Log-Likelihood:                 19.790
No. Observations:                  11   AIC:                            -35.58
Df Residuals:                       9   BIC:                            -34.78
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          4.1328      0.481      8.590      0.0

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year Five

In [47]:
x = df['div_yield'].iloc[147:158].tolist()
y = df['lg_excess_return'].iloc[147:158].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.249
Model:                            OLS   Adj. R-squared:                  0.166
Method:                 Least Squares   F-statistic:                     2.984
Date:                Thu, 27 Oct 2022   Prob (F-statistic):              0.118
Time:                        17:27:58   Log-Likelihood:                 17.115
No. Observations:                  11   AIC:                            -30.23
Df Residuals:                       9   BIC:                            -29.43
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          2.7641      0.387      7.139      0.0

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year Seven

In [48]:
x = df['div_yield'].iloc[171:182].tolist()
y = df['lg_excess_return'].iloc[171:182].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.281
Model:                            OLS   Adj. R-squared:                  0.201
Method:                 Least Squares   F-statistic:                     3.513
Date:                Thu, 27 Oct 2022   Prob (F-statistic):             0.0936
Time:                        17:28:31   Log-Likelihood:                 7.8322
No. Observations:                  11   AIC:                            -11.66
Df Residuals:                       9   BIC:                            -10.87
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          3.5283      0.499      7.072      0.0

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year Ten

In [49]:
x = df['div_yield'].iloc[207:218].tolist()
y = df['lg_excess_return'].iloc[207:218].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.120
Model:                            OLS   Adj. R-squared:                  0.022
Method:                 Least Squares   F-statistic:                     1.224
Date:                Thu, 27 Oct 2022   Prob (F-statistic):              0.297
Time:                        17:32:16   Log-Likelihood:                -2.2661
No. Observations:                  11   AIC:                             8.532
Df Residuals:                       9   BIC:                             9.328
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          1.9843      1.639      1.211      0.2

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Conclusion
    Year1:
$$
    R_1 = 4.556 - 26.1949\ (\frac{D}{P})_1\\
    R^2 = 0.886,\ 
    t-stats = -8.353
$$
    
    Year 2:
$$
    R_2 = 5.3309 - 33.864\ (\frac{D}{P})_2\\
    R^2 = 0.163,\ 
    t-stats = -1.324
$$

    Year 3:
$$
    R_3 = 4.1328 - 11.8674\ (\frac{D}{P})_3\\
    R^2 = 0.2,\ 
    t-stats = -1.498
$$

    Year 5:
$$
    R_5 = 2.7641 + 11.4694\ (\frac{D}{P})_5\\
    R^2 = 0.249,\ 
    t-stats = 1.727
$$

    Year 7:
$$
    R_7 = 3.528 + 20.2782\ (\frac{D}{P})_7\\
    R^2 = 0.281,\ 
    t-stats = 1.874
$$

    Year 10:
$$
    R_{10} = 1.9843 + 144.9267\ (\frac{D}{P})_{10}\\
    R^2 = 0.12,\ 
    t-stats = 1.106
$$

    From the above OLS regression, we can see that for year 1 to year 3 dividend yields have negative correlation with real excess return. The negative relationship indicating when price increase, real return will increase as well. However, for the following years, dividend yields have positive correlation with real excess return. The positive correlation indicates when price goes up, real return will go down. 

## Regress $ d_{t,\ t+k} $ 

In [85]:
df['real_growth_rate_div'] = (np.log(df['real_div'])).pct_change()

### Year 1:

In [86]:
x = df['div_yield'].iloc[99:110].tolist()
y = df['real_growth_rate_div'].iloc[99:110].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.202
Model:                            OLS   Adj. R-squared:                  0.113
Method:                 Least Squares   F-statistic:                     2.274
Date:                Thu, 27 Oct 2022   Prob (F-statistic):              0.166
Time:                        20:25:35   Log-Likelihood:                 60.134
No. Observations:                  11   AIC:                            -116.3
Df Residuals:                       9   BIC:                            -115.5
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0068      0.006     -1.175      0.2

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year 2:


In [87]:
x = df['div_yield'].iloc[111:122].tolist()
y = df['real_growth_rate_div'].iloc[111:122].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.000
Model:                            OLS   Adj. R-squared:                 -0.111
Method:                 Least Squares   F-statistic:                  0.002897
Date:                Thu, 27 Oct 2022   Prob (F-statistic):              0.958
Time:                        20:25:42   Log-Likelihood:                 61.901
No. Observations:                  11   AIC:                            -119.8
Df Residuals:                       9   BIC:                            -119.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0063      0.012      0.530      0.6

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year 3:

In [88]:
x = df['div_yield'].iloc[123:134].tolist()
y = df['real_growth_rate_div'].iloc[123:134].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.016
Model:                            OLS   Adj. R-squared:                 -0.094
Method:                 Least Squares   F-statistic:                    0.1423
Date:                Thu, 27 Oct 2022   Prob (F-statistic):              0.715
Time:                        20:25:43   Log-Likelihood:                 66.723
No. Observations:                  11   AIC:                            -129.4
Df Residuals:                       9   BIC:                            -128.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0088      0.007      1.310      0.2

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year 5:

In [89]:
x = df['div_yield'].iloc[147:158].tolist()
y = df['real_growth_rate_div'].iloc[147:158].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.580
Model:                            OLS   Adj. R-squared:                  0.533
Method:                 Least Squares   F-statistic:                     12.43
Date:                Thu, 27 Oct 2022   Prob (F-statistic):            0.00645
Time:                        20:25:43   Log-Likelihood:                 64.400
No. Observations:                  11   AIC:                            -124.8
Df Residuals:                       9   BIC:                            -124.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0165      0.005      3.135      0.0

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year 7:

In [90]:
x = df['div_yield'].iloc[171:182].tolist()
y = df['real_growth_rate_div'].iloc[171:182].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.749
Model:                            OLS   Adj. R-squared:                  0.721
Method:                 Least Squares   F-statistic:                     26.87
Date:                Thu, 27 Oct 2022   Prob (F-statistic):           0.000577
Time:                        20:25:44   Log-Likelihood:                 74.263
No. Observations:                  11   AIC:                            -144.5
Df Residuals:                       9   BIC:                            -143.7
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const          0.0074      0.001      6.199      0.0

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Year 10:

In [91]:
x = df['div_yield'].iloc[207:218].tolist()
y = df['real_growth_rate_div'].iloc[207:218].tolist()
x = sm.add_constant(x)
result = sm.OLS(y,x).fit()
print(result.summary())

                            OLS Regression Results                            
Dep. Variable:                      y   R-squared:                       0.260
Model:                            OLS   Adj. R-squared:                  0.177
Method:                 Least Squares   F-statistic:                     3.155
Date:                Thu, 27 Oct 2022   Prob (F-statistic):              0.109
Time:                        20:25:44   Log-Likelihood:                 64.706
No. Observations:                  11   AIC:                            -125.4
Df Residuals:                       9   BIC:                            -124.6
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const         -0.0080      0.004     -2.164      0.0

/Users/fengyujie/opt/anaconda3/lib/python3.9/site-packages/scipy/stats/stats.py:1541: UserWarning: kurtosistest only valid for n>=20 ... continuing anyway, n=11
  warnings.warn("kurtosistest only valid for n>=20 ... continuing "


### Conclusion:

    Year 1:
$$
    d_1 = -0.0068 + 0.1629\ (\frac{D}{P})_1\\
    R^2 = 0.202,\ t-stats = 1.508
$$
    
    Year 2:
$$
    d_2 = 0.0063 + 0.0103\ (\frac{D}{P})_2\\
    R^2 = 0,\ t-stats = 0.054
$$

    Year 3:
$$
    d_3 = 0.0088 - 0.0419\ (\frac{D}{P})_3\\
    R^2 = 0.016,\ t-stats = -0.377
$$

    Year 5:
$$
    d_5 = 0.0165 - 0.3181\ (\frac{D}{P})_5\\
    R^2 = 0.58,\ t-stats = -3.526
$$

    Year 7:
$$
    d_7 = 0.0074 - 0.1337\ (\frac{D}{P})_7\\
    R^2 = 0.749,\ t-stats = -5.184
$$

    Year 10:
$$
    d_{10} = -0.008 + 0.1637\ (\frac{D}{P})_{10}\\
    R^2 = 0.26,\ t-stats = 1.1776
$$

    In year 1, 2 and 3, log real growth rate of dividends has a positive relationship with dividend yield. However, in year 2, the R^2 is 0; therefore, the result from year 2 is invalid. The positive relationship of y and x indicates when price increases, real growth rate of dividend will decrease. 
    In year 3, 5 and 7, log real growth rate of dividends has negative relationship with dividend yield. The negative realtionship indicates when price increases, real growth rate of dividend will decrease. 
    In theory, when 